### DISCUSSION TITLE
preprocessor_error with msvc

### DISCUSSION THREAD

**Discussion by LinzzMichael** [ccache/ccache#1319] [Support]
### How to reproduce ###
config my project follow [this](https://github.com/ccache/ccache/wiki/MS-Visual-Studio#installation)

### Actual behavior ###
I found no cache file generate in the cache dir, every file output the error

```
Result: direct_cache_miss
Result: local_storage_read_miss
Result: preprocessor_error
```

In the debug log, it shows:
```
[2023-08-02T13:18:10.558356 38752] Preprocessor gave exit status 2
[2023-08-02T13:18:10.558392 38752] Failed; falling back to running the real compiler
```

### Expected behavior ###
it seems the debug log didn't show what cause the error, how can I slove this problem?

### Environment ###

ccache version 4.8.2
msvc 14.2
vs2019

**Comment by jrosdahl** [comment]
1. Locate `Running preprocessor` in the log and then the following `Executing ....` log line. It will look something like this:

    ```
    /path/to/compiler/cl.exe -P -Fi/path/to/tmp/dir/cpp_stdout.tmp.SA1BOK.i file.c
    ```

2. Run the command yourself without the `-Fi` part. Example:

    ```
    /path/to/compiler/cl.exe -P file.c
    ```

3. Check the output for clues.

**Reply by LinzzMichael** [reply]
hank you for your help, his method works。but I found another problem
when working with msvc and /yu, the cache always miss hit, In the debug log, it shows:

```
[2023-08-10T14:17:09.524959 55244] Result: cache_miss
[2023-08-10T14:17:09.524962 55244] Result: direct_cache_miss
[2023-08-10T14:17:09.524963 55244] Result: local_storage_miss
[2023-08-10T14:17:09.524964 55244] Result: local_storage_read_hit
[2023-08-10T14:17:09.524965 55244] Result: local_storage_read_miss
[2023-08-10T14:17:09.524966 55244] Result: local_storage_write
[2023-08-10T14:17:09.524970 55244] Result: local_storage_write
[2023-08-10T14:17:09.524971 55244] Result: preprocessed_cache_miss
```


does ccache can not work with msvc and precompiler header? @jrosdahl

**Reply by cristianadam** [reply]
For cl (and clang-cl) ccache should store two artifacts `.pch` and `.obj` files for precomipled headers. Currently for gcc/clang this is not the case there is only one output artifact: the `.pch` / `.gch` precompiled file.

If ccache could store two artifacts, it should be able to support precompiled headers.

**Reply by jrosdahl** [reply]
> does ccache can not work with msvc and precompiler header? @jrosdahl

AFAIK it should work with the caveat mentioned in https://github.com/ccache/ccache/pull/991#issue-1098801109. @orgads: Any input here?

**Reply by orgads** [reply]
Not sure. I never tried it myself :)

Maybe @llunak can help.
